keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,219 @@
+trying to recovery from the failed node. The recovery process will taint the node
+status as “NotReady” and will trigger eviction of pods on the failed node if the node
+is in "NotReady" state for a while (set to 5 minutes by default) so that new pods
+could be started on other healthy nodes. 
+
+However, users often reported issues during the failure handling including old pods
+cannot be deleted correctly, new pods cannot be started correctly due to volume failing
+to detach from old node and attach to new node issues, or it takes a long time for new
+pods to start. The main reason for this problem is due to the fact that Pod deletion requires
+kubelet's acknowledgement. When a Pod is evicted (deleted) from a node, the deletion timestamp
+is first set on the Pod. However the Pod will not be actually deleted from API server until kubelet 
+finishes cleaning up all the resources of this Pod and kills the running containers. If the node
+is in a failed state and kubelet cannot respond, Pod will be stuck in a terminating state forever.
+In addition, the containers status for the Pod are still running since kubelet cannot update it,
+which blocks attach_detach controller to detach volumes. The current taint-based eviction actually
+cannot give any benefit in the sense that Pod can only be successfully evicted after Kubelet comes
+back healthy, in which case there is no need of Pod eviction at all.
+
+
+### Goals
+
+This proposal is trying to investigate the failure handling issues for pods
+and their volumes with a goal of improving system reliability and failure recovery capabilities.
+Among different kinds of failures, node shut down is the main user case this proposal tries
+to address. In this case, we are sure that all containers are stopped, and all mounts are gone
+so that it is safe to take some recovery actions more aggressively without waiting for kubelet correspondence.  
+
+to delete the Pods from API server without waiting for kubelet's acknowledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+### Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects: 
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+### ForceEviction Taint Effect
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
```

Comment:
It's not clear to me why a new taint effect is required.

This document describes the following interlock:
1. controller gets positive confirmation of shut down node state, fetches Node object, verifies node is NotReady, updates with taint (which would fail with a conflict if the node was updated in the meantime), verifies node is NotReady and tainted, then evicts pods
2. kubelet reports Ready on startup, starts honoring taint when choosing whether to start pods or not

The novel aspect of this proposal seems to be to add taint-interpreting logic into node admission. If the interlock works as described, is there a reason a designated taint (e.g. `Shutdown`) with a NoExecute effect is insufficient?

(note also that any feature depending on an interlock with kubelets has a 3 release period from the point it is enabled by default before the control plane can assume its presence)

Reply:
Do you mean we still use "NoExecute" taint effect, but give it a new interpretation to allow pods to be deleted forcefully with 0 grace period as we proposed for "ForceEviction" taint.

We discussed this option a little bit, the big concern I have is it will change the current behavior of NoExecute taint effect. Controller or system admin might taint the node with "NoExecute" effect for different purposes. E.g., to evict pods for node maintenance or upgrade. So in these cases, node is still running and kubelet is also up, pods can be evicted successfully without setting force option and 0 grace period.

NoExecute effect is not working for shutdown case because NoExecute evicts Pods without force option.
1. Node shut down
2. Node becomes NotReady and taint with NoExecute effect
3. Taint manager tries to evict pods.
4. Pods are stuck in terminating states as long as node is shut down because without kubelet acknowledgement, pods cannot be deleted from API server. If pods belong to a statefulset, no new pods can be started since the old one can not be deleted.
5. Assume node is restarted after a while, pods can be deleted finally after kubelet restarts and deleted those Pods forcefully. But there is almost no benefit to evict those pods at this point since kubelet comes back.

With this new taint effect, we basically specifying the eviction is a force eviction instead of a soft one. I don't know if there is any other way to give the same taint effect with different options/behavior.

Reply:
I think NoExecute is orthogonal to pod safety. A NoExecute taint means "remove pods that don't tolerate this taint, after time period X, safely".

If we can design an interlock between the controller, cloud provider API, and kubelet so that we can delete these pods safely even in the presence of a kubelet that cannot ack the removal because it is shut down, that seems preferable to introducing a new taint effect.

Reply:
how about bare-metal case with no cloud provider API? I think it is the main use case we are targeting.